### DISCUSSION TITLE
Creating Trial- and Block-indices when using Timeline Variables

### DISCUSSION THREAD

**Discussion by Raphael-Merz** [jspsych/jsPsych#2928] [Q&A] [answered]
Hi everyone,

As a short disclaimer: I'm relatively new to jsPsych (and javascript) and have only just been trying out a couple of things these past few months!
In my experiment, I have participants repeat the same task a bunch of times. The task itself consists of different steps that work just fine. To randomize the order of stimuli, I've been using Timeline Variables and the designated functions:

```
`
    var together_encoding = {  // Encoding-Task
        timeline: [first_fixation_of_a_trial, task_part_1, fixation, task_part_2, task_part_3, fixation, task_part_4],
        timeline_variables: task_words_encoding,
        sample: {
            type: 'without-replacement',
            size: 15
        },
        randomize_order: true
            };
    timeline.push(together_encoding);
`
```

Now to my question: In the final data, I want to have one continuous Trial-Id that only changes when a new trial begins (i.e., all steps within the timeline above have the same value until a repetition starts). I've seen some code online that used the loop_function, but I wasn't 100% sure if that'd work here, since I'm basically looping using Timeline Variables above. I also tried to create some variable that'd change with each "first_fixation_of_a_trial" but that didn't work either.

If somebody could refer me to a similar discussion, or knows an easy fix, that'd be great!

Thanks so much and take care
Raphael

**Comment by jodeleeuw** [comment, marked as answer]
Hi @Raphael-Merz,

>  I also tried to create some variable that'd change with each "first_fixation_of_a_trial" but that didn't work either.

I think this is the approach that I would take. It is a bit tricky to implement because you have to use a [dynamic parameter](https://www.jspsych.org/7.3/overview/dynamic-parameters/).

Here's a simplified example:

```js
var block_index = -1; // starting this at -1 so the first block is 0 after incrementing.

var first_trial = {
	// whatever parameters you need here
	on_start: function(){
		block_index++;
	}
}

var block_timeline = {
	timeline: [ first_trial, second_trial, third_trial ],
	data: {
		block_id: function() { return block_index; }
	}
}
```

This should add `block_id` to every trial in the `block_timeline`, and use the value of `block_index`. This value will get updated when `first_trial` starts.

**Reply by Raphael-Merz** [reply]
Hi @jodeleeuw,
Works very well now, thank you very much!
